charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,207 @@
+    # setup Istio
+    istio_gateway = "istio-ingressgateway"
+    istio_pilot = "istio-pilot"
+    await _setup_istio(ops_test, istio_gateway, istio_pilot)
+
+    # add Tensorboard-Controller/Istio relation
+    await ops_test.model.add_relation(f"{istio_pilot}:gateway-info", f"{APP_NAME}:gateway-info")
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=60 * 5
+    )
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_replicas(
+    client: Client,
+    resource_class: Resource,
+    resource_name: str,
+    namespace: str,
+    target_replicas: int = 1,
+):
+    """Test for replicas.
+
+    Retries multiple times to allow for K8s resource to reach the target number of ready replicas.
+    """
+    rsc = client.get(resource_class, resource_name, namespace=namespace)
+    replicas = rsc.get("status", {}).get("readyReplicas")
+
+    resource = f"{namespace}/{resource_class.kind}/{resource_name}"
+    logger.info(f"Waiting for {resource} to reach {target_replicas} ready replicas:"
+                f" readyReplicas == {replicas}")
+
+    assert replicas == target_replicas, f"Waited too long for {resource}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # manually delete the PVC, since Juju is not aware of it
+    lightkube_client.delete(PersistentVolumeClaim, name=PVC_NAME, namespace=ops_test.model_name)
```

Comment:
This stands out as a bit odd.  We remove this object here, then in a few lines test if it is removed.  We're effectively just testing that our test worked.

I can see the reason for removing this manually so we clean up our test environment, but it feels odd here.  If anything, removing the PVC and Tensorboard CR feels more like a cleanup step from the previous test that would work well in a fixture